File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Cdm/Projections/CdmOperationReplaceAsForeignKey.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Cdm
{
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.ResolvedModel;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;
    using System;
    using System.Collections.Generic;
    using System.Linq;

    /// <summary>
    /// Class to handle ReplaceAsForeignKey operations
    /// </summary>
    public class CdmOperationReplaceAsForeignKey : CdmOperationBase
    {
        private static readonly string Tag = nameof(CdmOperationReplaceAsForeignKey);

        public string Reference { get; set; }

        public CdmTypeAttributeDefinition ReplaceWith { get; set; }

        public CdmOperationReplaceAsForeignKey(CdmCorpusContext ctx) : base(ctx)
        {
            this.ObjectType = CdmObjectType.OperationReplaceAsForeignKeyDef;
            this.Type = CdmOperationType.ReplaceAsForeignKey;
        }

        /// <inheritdoc />
        public override CdmObject Copy(ResolveOptions resOpt = null, CdmObject host = null)
        {
            if (resOpt == null)
            {
                resOpt = new ResolveOptions(this, this.Ctx.Corpus.DefaultResolutionDirectives);
            }

            var copy = host == null ?  new CdmOperationReplaceAsForeignKey(this.Ctx) : host as CdmOperationReplaceAsForeignKey;

            copy.ReplaceWith = this.ReplaceWith?.Copy(resOpt) as CdmTypeAttributeDefinition;
            copy.Reference = this.Reference;

            this.CopyProj(resOpt, copy);
            return copy;
        }

        /// <inheritdoc />
        [Obsolete("CopyData is deprecated. Please use the Persistence Layer instead.")]
        public override dynamic CopyData(ResolveOptions resOpt = null, CopyOptions options = null)
        {
            return CdmObjectBase.CopyData<CdmOperationReplaceAsForeignKey>(this, resOpt, options);
        }

        public override string GetName()
        {
            return "operationReplaceAsForeignKey";
        }

        /// <inheritdoc />
        [Obsolete]
        public override CdmObjectType GetObjectType()
        {
            return CdmObjectType.OperationReplaceAsForeignKeyDef;
        }

        /// <inheritdoc />
        public override bool Validate()
        {
            List<string> missingFields = new List<string>();

            if (string.IsNullOrWhiteSpace(this.Reference))
                missingFields.Add("Reference");

            if (this.ReplaceWith == null)
                missingFields.Add("ReplaceWith");

            if (missingFields.Count > 0)
            {
                Logger.Error(this.Ctx, Tag, nameof(Validate), this.AtCorpusPath, CdmLogCode.ErrValdnIntegrityCheckFailure, this.AtCorpusPath, string.Join(", ", missingFields.Select((s) =>$"'{s}'")));
                return false;
            }

            return true;
        }

        /// <inheritdoc />
        public override bool Visit(string pathFrom, VisitCallback preChildren, VisitCallback postChildren)
        {
            string path = this.UpdateDeclaredPath(pathFrom);

            if (preChildren != null && preChildren.Invoke(this, path))
                return false;

            if (this.ReplaceWith != null)
            {
                if (this.ReplaceWith.Visit(pathFrom + "foreignKeyAttribute/", preChildren, postChildren))
                {
                    return true;
                }
            }

            if (postChildren != null && postChildren.Invoke(this, path))
                return true;

            return false;
        }

        /// <inheritdoc />
        internal override ProjectionAttributeStateSet AppendProjectionAttributeState(
            ProjectionContext projCtx,
            ProjectionAttributeStateSet projOutputSet,
            CdmAttributeContext attrCtx)
        {
            // Create new attribute context for the operation
            AttributeContextParameters attrCtxOpFKParam = new AttributeContextParameters
            {
                under = attrCtx,
                type = CdmAttributeContextType.OperationReplaceAsForeignKey,
                Name = $"operation/index{Index}/operationReplaceAsForeignKey"
            };
            CdmAttributeContext attrCtxOpFK = CdmAttributeContext.CreateChildUnder(projCtx.ProjectionDirective.ResOpt, attrCtxOpFKParam);

            // Create new attribute context for the AddedAttributeIdentity
            AttributeContextParameters attrCtxFKParam = new AttributeContextParameters
            {
                under = attrCtxOpFK,
                type = CdmAttributeContextType.AddedAttributeIdentity,
                Name = $"_foreignKey"
            };
            CdmAttributeContext attrCtxFK = CdmAttributeContext.CreateChildUnder(projCtx.ProjectionDirective.ResOpt, attrCtxFKParam);

            // get the added attribute and applied trait
            CdmTypeAttributeDefinition subFK = this.ReplaceWith;
            List<string> addTrait = new List<string>() { "is.linkedEntity.identifier" };

            // Create new resolved attribute, set the new attribute as target, and apply "is.linkedEntity.identifier" trait
            ResolvedAttribute resAttrNewFK = CreateNewResolvedAttribute(projCtx, attrCtxFK, subFK, addedSimpleRefTraits: addTrait);

            ProjectionAttributeStateSet outputFromOpPasSet = CreateNewProjectionAttributeStateSet(projCtx, projOutputSet, resAttrNewFK, this.Reference);

            return outputFromOpPasSet;
        }

        private static ProjectionAttributeStateSet CreateNewProjectionAttributeStateSet(
            ProjectionContext projCtx,
            ProjectionAttributeStateSet projOutputSet,
            ResolvedAttribute newResAttrFK,
            string refAttrName)
        {
            List<ProjectionAttributeState> pasList = ProjectionResolutionCommonUtil.GetLeafList(projCtx, refAttrName);
            string sourceEntity = projCtx.ProjectionDirective.OriginalSourceAttributeName;

            if (sourceEntity == null)
            {
                Logger.Warning(projOutputSet.Ctx, Tag, nameof(CreateNewProjectionAttributeStateSet), null, CdmLogCode.WarnProjFKWithoutSourceEntity, refAttrName);
            }

            if (pasList != null)
            {
                // update the new foreign key resolved attribute with trait param with reference details
                ResolvedTrait reqdTrait = newResAttrFK.ResolvedTraits.Find(projCtx.ProjectionDirective.ResOpt, "is.linkedEntity.identifier");
                if (reqdTrait != null && sourceEntity != null)
                {
                    CdmEntityReference traitParamEntRef = ProjectionResolutionCommonUtil.CreateForeignKeyLinkedEntityIdentifierTraitParameter(projCtx.ProjectionDirective, projOutputSet.Ctx.Corpus, pasList);
                    reqdTrait.ParameterValues.SetParameterValue(projCtx.ProjectionDirective.ResOpt, "entityReferences", traitParamEntRef);
                }

                // Create new output projection attribute state set for FK and add prevPas as previous state set
                ProjectionAttributeState newProjAttrStateFK = new ProjectionAttributeState(projOutputSet.Ctx)
                {
                    CurrentResolvedAttribute = newResAttrFK,
                    PreviousStateList = pasList
                };

                projOutputSet.Add(newProjAttrStateFK);
            }
            else
            {
                // Log error & return projOutputSet without any change
                Logger.Error(projOutputSet.Ctx, Tag, nameof(CreateNewProjectionAttributeStateSet), null, CdmLogCode.ErrProjRefAttrStateFailure, refAttrName);
            }

            return projOutputSet;
        }
    }
}

File: objectModel/TypeScript/Cdm/Projections/CdmOperationReplaceAsForeignKey.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import {
    CdmAttributeContext,
    cdmAttributeContextType,
    CdmCorpusContext,
    CdmEntityReference,
    CdmObject,
    cdmObjectType,
    CdmOperationBase,
    cdmOperationType,
    CdmTypeAttributeDefinition,
    cdmLogCode,
    Logger,
    ProjectionAttributeState,
    ProjectionAttributeStateSet,
    ProjectionContext,
    ProjectionResolutionCommonUtil,
    ResolvedAttribute,
    ResolvedTrait,
    resolveOptions,
    VisitCallback,
    StringUtils
} from '../../internal';
import { AttributeContextParameters } from '../../Utilities/AttributeContextParameters';

/**
 * Class to handle ReplaceAsForeignKey operations
 */
export class CdmOperationReplaceAsForeignKey extends CdmOperationBase {
    private TAG: string = CdmOperationReplaceAsForeignKey.name;
    public reference: string;
    public replaceWith: CdmTypeAttributeDefinition;

    constructor(ctx: CdmCorpusContext) {
        super(ctx);
        this.objectType = cdmObjectType.operationReplaceAsForeignKeyDef;
        this.type = cdmOperationType.replaceAsForeignKey;
    }

    /**
     * @inheritdoc
     */
     public copy(resOpt?: resolveOptions, host?: CdmObject): CdmObject {
        if (!resOpt) {
            resOpt = new resolveOptions(this, this.ctx.corpus.defaultResolutionDirectives);
        }

        const copy: CdmOperationReplaceAsForeignKey = !host ? new CdmOperationReplaceAsForeignKey(this.ctx) : host as CdmOperationReplaceAsForeignKey;

        copy.replaceWith = this.replaceWith ? this.replaceWith.copy(resOpt) as CdmTypeAttributeDefinition : undefined;
        copy.reference = this.reference;
        
        this.copyProj(resOpt, copy);
        return copy;
    }

    /**
     * @inheritdoc
     */
    public getName(): string {
        return 'operationReplaceAsForeignKey';
    }

    /**
     * @inheritdoc
     */
    public getObjectType(): cdmObjectType {
        return cdmObjectType.operationReplaceAsForeignKeyDef;
    }

    /**
     * @inheritdoc
     */
    public validate(): boolean {
        const missingFields: string[] = [];

        if (!this.reference) {
            missingFields.push('reference');
        }

        if (!this.replaceWith) {
            missingFields.push('replaceWith');
        }

        if (missingFields.length > 0) {
            Logger.error(this.ctx, this.TAG, this.validate.name, this.atCorpusPath, cdmLogCode.ErrValdnIntegrityCheckFailure, missingFields.map((s: string) => `'${s}'`).join(', '));
            return false;
        }

        return true;
    }

    /**
     * @inheritdoc
     */
    public visit(pathFrom: string, preChildren: VisitCallback, postChildren: VisitCallback): boolean {
        let path: string = '';
        if (!this.ctx.corpus.blockDeclaredPathChanges) {
            path = this.declaredPath;
            if (!path) {
                path = pathFrom + 'operationReplaceAsForeignKey';
                this.declaredPath = path;
            }
        }

        if (preChildren && preChildren(this, path)) {
            return false;
        }

        if (this.replaceWith) {
            if (this.replaceWith.visit(pathFrom + 'foreignKeyAttribute/', preChildren, postChildren)) {
                return true;
            }
        }

        if (postChildren && postChildren(this, path)) {
            return true;
        }

        return false;
    }

    /**
     * @inheritdoc
     * @internal
     */
    public appendProjectionAttributeState(
        projCtx: ProjectionContext,
        projOutputSet: ProjectionAttributeStateSet,
        attrCtx: CdmAttributeContext
    ): ProjectionAttributeStateSet {
        // Create new attribute context for the operation
        const attrCtxOpFKParam: AttributeContextParameters = {
            under: attrCtx,
            type: cdmAttributeContextType.operationReplaceAsForeignKey,
            name: `operation/index${this.index}/operationReplaceAsForeignKey`
        };
        const attrCtxOpFK: CdmAttributeContext = CdmAttributeContext.createChildUnder(projCtx.projectionDirective.resOpt, attrCtxOpFKParam);

        // Create new attribute context for the AddedAttributeIdentity
        const attrCtxFKParam: AttributeContextParameters  = {
            under: attrCtxOpFK,
            type: cdmAttributeContextType.addedAttributeIdentity,
            name: '_foreignKey'
        };
        const attrCtxFK: CdmAttributeContext = CdmAttributeContext.createChildUnder(projCtx.projectionDirective.resOpt, attrCtxFKParam);

        // get the added attribute and applied trait
        const subFK: CdmTypeAttributeDefinition = this.replaceWith;
        const addTrait: string[] = ['is.linkedEntity.identifier'];

        // Create new resolved attribute, set the new attribute as target, and apply "is.linkedEntity.identifier" trait
        const resAttrNewFK: ResolvedAttribute = CdmOperationBase.createNewResolvedAttribute(projCtx, attrCtxFK, subFK, undefined, addTrait);

        const outputFromOpPasSet: ProjectionAttributeStateSet = CdmOperationReplaceAsForeignKey.createNewProjectionAttributeStateSet(projCtx, projOutputSet, resAttrNewFK, this.reference);

        return outputFromOpPasSet;
    }

    private static createNewProjectionAttributeStateSet(
        projCtx: ProjectionContext,
        projOutputSet: ProjectionAttributeStateSet,
        newResAttrFK: ResolvedAttribute,
        refAttrName: string
    ): ProjectionAttributeStateSet {
        const pasList: ProjectionAttributeState[] = ProjectionResolutionCommonUtil.getLeafList(projCtx, refAttrName);
        const sourceEntity = projCtx.projectionDirective.originalSourceEntityAttributeName;

        if (!sourceEntity) {
            Logger.warning(projOutputSet.ctx, CdmOperationReplaceAsForeignKey.name, this.createNewProjectionAttributeStateSet.name, null, cdmLogCode.WarnProjFKWithoutSourceEntity, refAttrName);
        }

        if (pasList) {
            // update the new foreign key resolved attribute with trait param with reference details
            const reqdTrait: ResolvedTrait = newResAttrFK.resolvedTraits.find(projCtx.projectionDirective.resOpt, 'is.linkedEntity.identifier');
            if (reqdTrait && sourceEntity) {
                const traitParamEntRef: CdmEntityReference = ProjectionResolutionCommonUtil.createForeignKeyLinkedEntityIdentifierTraitParameter(projCtx.projectionDirective, projOutputSet.ctx.corpus, pasList);
                reqdTrait.parameterValues.setParameterValue(projCtx.projectionDirective.resOpt, 'entityReferences', traitParamEntRef);
            }

            // Create new output projection attribute state set for FK and add prevPas as previous state set
            const newProjAttrStateFK: ProjectionAttributeState = new ProjectionAttributeState(projOutputSet.ctx);
            newProjAttrStateFK.currentResolvedAttribute = newResAttrFK;
            newProjAttrStateFK.previousStateList = pasList;

            projOutputSet.add(newProjAttrStateFK);
        } else {
            // Log error & return projOutputSet without any change
            Logger.error(projOutputSet.ctx, CdmOperationReplaceAsForeignKey.name, this.createNewProjectionAttributeStateSet.name, null, cdmLogCode.ErrProjRefAttrStateFailure, refAttrName);
        }

        return projOutputSet;
    }
}
